### PR TITLE
reindex delete batches

### PR DESCRIPTION
**Pull request by geopin90**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -128,6 +129,19 @@ async fn reindex(
         )?;
       }
 
+      // delete batch from document if it exists
+      let mut after = if after["batch"].is_object() {
```
Comment (shabanovd) [inline src/main.rs:133]: why `after.remove("batch");` can't be used?

File: src/storage/mod.rs [hunk 2]
```diff
@@ -49,3 +49,13 @@ pub(crate) fn save(path: &PathBuf, data: String) -> Result<(), Error> {
     .write_all(data.as_bytes())
     .map_err(|e| Error::IOError(format!("fail to write file: {}", e)))
 }
+
+pub(crate) fn remove_dir(path: &PathBuf) -> Result<(), Error> {
```
Comment (shabanovd) [inline src/storage/mod.rs:53]: this is wrong because method delete parent folder, not provided one

File: store/src/elements.rs [hunk 3]
```diff
@@ -456,11 +402,11 @@ fn json_to_ops(
     }
   };
 
-  data["batch"] = object! {
-    "_uuid": batch.id.to_string(),
-    "date": time_to_string(batch.date),
-    "barcode": batch.to_barcode(),
-  };
+  // data["batch"] = object! {
```
Comment (shabanovd) [inline store/src/elements.rs:405]: delete this, and make `data: &mut JsonValue,` immutable

File: src/storage/mod.rs (hunk 2 above)
Comment (geopin90) [reply src/storage/mod.rs:53]: corrected to path.as_path() instead of path.parent()

File: store/src/elements.rs (hunk 3 above)
Comment (shabanovd) [reply store/src/elements.rs]: it must not happen here

File: src/main.rs [hunk 4]
```diff
@@ -28,6 +28,7 @@ use actix_web_httpauth::extractors::bearer::{BearerAuth, Config};
 use actix_web_httpauth::extractors::AuthenticationError;
 use actix_web_httpauth::middleware::HttpAuthentication;
 
+use config::FileFormat::Json;
```
Comment (shabanovd) [inline src/main.rs:31]: why you need it here?

File: src/use_cases/uc_save.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+use json::object;
+use service::utils::json::JsonParams;
+use service::{Context, Services};
+use std::fs::{File, OpenOptions};
+use std::io::{BufRead, BufReader, Error, ErrorKind};
+
+pub fn save(app: &Application) -> Result<(), Error> {
+  let mut count = 0;
+
+  for ws in app.wss.list()? {
```
Comment (shabanovd) [inline src/use_cases/uc_save.rs:13]: why you can't directly call required workspace?

File: store/src/elements.rs [hunk 6]
```diff
@@ -511,7 +534,14 @@ fn goods(
         };
 
       if let Some(goods) = product["goods"].string_or_none() {
-        Ok(app.service("memories").get(Context::local(), goods, goods_params)?)
+        let goods = app.service("memories").get(Context::local(), goods, goods_params)?;
+        write_to_file(
```
[the remark this answers is not in the document]
Comment (geopin90) [reply store/src/elements.rs]: do it in uc_save now

File: src/storage/memories.rs [hunk 7]
```diff
@@ -72,7 +72,9 @@ fn save_data(
     Ok(b) => {
       //WORKAROUND: make sure that id & uuid stay same
       data["_id"] = b["_id"].clone();
-      data["_uuid"] = b["_uuid"].clone();
+      if !b["_uuid"].is_null() {
```
Comment (shabanovd) [inline src/storage/memories.rs:75]: same check must be done on `b["_id"]` too

File: store/src/elements.rs [hunk 8]
```diff
@@ -536,45 +536,46 @@ fn goods(
...
-              // log::debug!("_new_ {res:?}");
-              Ok(res)
-            },
-            _ => Err(WHError::new("More than one goods with same name")),
-          }
-        } else {
-          Err(WHError::new("No data for goods"))
-        };
+        Err(WHError::new("No data for goods"))
+        // return if let (Some(customer), Some(label)) =
```
Comment (shabanovd) [inline store/src/elements.rs:540]: remove comments

File: src/use_cases/uc_delete.rs [hunk 9]
```diff
@@ -1,39 +1,60 @@
...
+use service::{Context, Services};
 use std::fs::File;
 use std::io::{BufRead, BufReader, Error};
 
 pub fn delete(app: &Application) -> Result<(), Error> {
   let mut count = 0;
-  for ws in app.wss.list()? {
-    for doc in ws.clone().into_iter() {
-      // println!("{:?} {:?}", doc.id, doc.json()?);
+  let ws = app.wss.list()?[0].clone();
```
Comment (shabanovd) [inline src/use_cases/uc_delete.rs:11]: why you think that `[0]` is correct one?

Comment (geopin90) [reply src/use_cases/uc_delete.rs:11]: Can do let ws = "yjmgJUmDo_kn9uxVi8s9Mj9mgGRJISxRt63wT46NyTQ" for Midas

File: src/use_cases/uc_delete.rs [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+use actix_web::web::patch;
+use json::object;
+use service::{Context, Services};
+use std::fs::File;
+use std::io::{BufRead, BufReader, Error};
+
+pub fn delete(app: &Application) -> Result<(), Error> {
+  let mut count = 0;
+  let ws = app.wss.list()?[0].clone();
+  for doc in ws.clone().into_iter() {
```
Comment (shabanovd) [inline src/use_cases/uc_delete.rs:12]: why you iterate over all documents?

File: src/use_cases/uc_delete.rs [hunk 11]
```diff
@@ -8,57 +8,59 @@ use std::io::{BufRead, BufReader, Error};
 
 pub fn delete(app: &Application) -> Result<(), Error> {
   let mut count = 0;
-  let ws = app.wss.list()?[0].clone();
-  for doc in ws.clone().into_iter() {
-    // println!("{:?} {:?}", doc.id, doc.json()?);
+  for ws in app.wss.list()? {
+    if ws.id.to_string() != "yjmgJUmDo_kn9uxVi8s9Mj9mgGRJISxRt63wT46NyTQ".to_string() {
```
Comment (shabanovd) [inline src/use_cases/uc_delete.rs:12]: why you iterate over all workspaces?

**Review by shabanovd — APPROVED** [review]

**Pull request merged**